src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1184,6 +1193,192 @@ void HNSWIndex<DataType, DistType>::resizeIndexInternal(size_t new_max_elements)
     max_elements_ = new_max_elements;
 }
 
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::mutuallyUpdateForRepairedNode(
+    idType node_id, size_t level, std::vector<idType> &neighbors_to_remove,
+    std::vector<idType> &nodes_to_update, std::vector<idType> &chosen_neighbors, size_t max_M_cur) {
+    // Sort the nodes to remove set for fast lookup.
+    std::sort(neighbors_to_remove.begin(), neighbors_to_remove.end());
+
+    // Acquire the required locks for the updates, after sorting the nodes to update
+    // (to avoid deadlocks)
+    nodes_to_update.push_back(node_id);
+    std::sort(nodes_to_update.begin(), nodes_to_update.end());
+    size_t nodes_to_update_count = nodes_to_update.size();
+    std::unique_lock<std::mutex> locks[nodes_to_update_count];
+    for (size_t i = 0; i < nodes_to_update_count; i++) {
+        locks[i] = std::unique_lock<std::mutex>(element_neighbors_locks_[nodes_to_update[i]]);
+    }
+
+    idType *node_neighbors = getNodeNeighborsAtLevel(node_id, level);
+    linkListSize node_neighbors_count = getNodeNeighborsCount(node_neighbors);
+    auto *node_incoming_edges = getIncomingEdgesPtr(node_id, level);
+
+    // Perform mutual updates: go over the node's neighbors and overwrite the neighbors to remove
+    // that are still exist.
+    size_t node_neighbors_idx = 0;
+    for (size_t i = 0; i < node_neighbors_count; i++) {
+        if (!std::binary_search(nodes_to_update.begin(), nodes_to_update.end(),
+                                node_neighbors[i])) {
+            // The repaired node added a new neighbor that we didn't account for before in the
+            // meantime - leave it as is.
+            node_neighbors[node_neighbors_idx++] = node_neighbors[i];
+            continue;
+        }
+        auto it = std::find(chosen_neighbors.begin(), chosen_neighbors.end(), node_neighbors[i]);
+        if (it != chosen_neighbors.end()) {
+            // A chosen neighbor is already connected to the node - leave it as is.
+            node_neighbors[node_neighbors_idx++] = node_neighbors[i];
+            chosen_neighbors.erase(it);
```

Comment:
1 - `it` is defined before just for the sake of readability (otherwise the condition in the `if` will be longer and messy... 
2 - right, will do

Reply:
I think it's better to well-scope this variable as good practice and to inform the compiler it won't be used again. but it's probably fine (scope ends few operations later, and maybe it can conclude it anyway)